drivers/motor_driver/motor_driver.c

```diff
@@ -29,155 +29,169 @@
-    uint16_t resol = motor_driver_conf->pwm_resolution;
+int motor_driver_init(motor_driver_t *motor_driver, const motor_driver_params_t *motor_driver_params)
+{
+    int err = 0;
 
-    uint32_t ret_pwm = pwm_init(pwm_dev, mode, freq, resol);
-    if (ret_pwm != freq) {
+    uint32_t ret_pwm = pwm_init(motor_driver_params->pwm_dev,
+                                motor_driver_params->pwm_mode,
+                                motor_driver_params->pwm_frequency,
+                                motor_driver_params->pwm_resolution);
+    if (ret_pwm != motor_driver_params->pwm_frequency) {
         err = EINVAL;
         LOG_ERROR("pwm_init failed\n");
         goto motor_init_err;
     }
 
-    for (uint8_t i = 0; i < motor_driver_conf->nb_motors; i++) {
-        if (gpio_is_valid(motor_driver_conf->motors[i].gpio_dir0)
-            && (gpio_init(motor_driver_conf->motors[i].gpio_dir0,
-                          GPIO_OUT))) {
-            err = EIO;
-            LOG_ERROR("gpio_dir0 init failed\n");
-            goto motor_init_err;
+    /* Init GPIO */
+    err = 0;
+    for (uint8_t i = 0; i < motor_driver_params->nb_motors; i++) {
+        /* Init motor GPIOs, if one fails, motor is not setup */
+        if (gpio_is_valid(motor_driver_params->motors[i].gpio_dir0)) {
+            if (gpio_init(motor_driver_params->motors[i].gpio_dir0,
+                          GPIO_OUT)) {
+                err = EIO;
+                LOG_ERROR("gpio_dir0 init failed for motor %d\n", i);
+                goto motor_init_err;
+            }
         }
-        if (gpio_is_valid(motor_driver_conf->motors[i].gpio_dir1_or_brake)
-            && (gpio_init(motor_driver_conf->motors[i].gpio_dir1_or_brake,
-                          GPIO_OUT))) {
-            err = EIO;
-            LOG_ERROR("gpio_dir1_or_brake init failed\n");
-            goto motor_init_err;
+        if (gpio_is_valid(motor_driver_params->motors[i].gpio_dir1_or_brake)) {
+            if (gpio_init(motor_driver_params->motors[i].gpio_dir1_or_brake,
+                          GPIO_OUT)) {
+                err = EIO;
+                LOG_ERROR("gpio_dir1_or_brake init failed for motor %d\n", i);
+                goto motor_init_err;
+            }
         }
-        if (gpio_is_valid(motor_driver_conf->motors[i].gpio_enable)) {
-            if (gpio_init(motor_driver_conf->motors[i].gpio_enable,
+        if (gpio_is_valid(motor_driver_params->motors[i].gpio_enable)) {
+            if (gpio_init(motor_driver_params->motors[i].gpio_enable,
                           GPIO_OUT)) {
                 err = EIO;
-                LOG_ERROR("gpio_enable init failed\n");
+                LOG_ERROR("gpio_enable init failed for motor %d\n", i);
                 goto motor_init_err;
             }
-            motor_enable(motor_driver, i);
         }
     }
 
-    return 0;
+    /* Set callbacks according to driver type */
+    switch(motor_driver_params->mode) {
```

Comment:
wouldn't it make more sense do a switch when the brake or direction function is called instead of storing an extra function pointer?

Reply:
The first idea was to make this two callbacks configurables with `motor_driver_params_t`. However as the driver is able to support most of existing motor drivers, I did not find really relevant to expose this callbacks inside `motor_driver_params_t`.
But I kept them internally, because it allows less lines of code at the price of 2 pointers, indeed.

As I do not have a real opinion about this, I follow your review. :wink: